hail/python/hailtop/aiogoogle/client/storage_client.py

```diff
@@ -128,6 +191,58 @@ async def create(self, url: str) -> WritableStream:
     async def mkdir(self, url: str) -> None:
         pass
 
+    async def statfile(self, url: str) -> GetObjectFileStatus:
+        try:
+            bucket, name = self._get_bucket_name(url)
+            return GetObjectFileStatus(await self._storage_client.get_object_metadata(bucket, name))
+        except aiohttp.ClientResponseError as e:
+            if e.status == 404:
+                raise FileNotFoundError(url)
+            raise
+
+    async def _listfiles_recursive(self, bucket: str, name: str) -> AsyncIterator[FileListEntry]:
+        params = {
+            'prefix': name
+        }
+        async for page in await self._storage_client.list_objects(bucket, params=params):
+            prefixes = page.get('prefixes')
+            assert not prefixes
+
+            items = page.get('items')
+            if items is not None:
+                for item in page['items']:
+                    yield GoogleStorageFileListEntry(f'gs://{bucket}/{item["name"]}', True, item)
+
+    async def _listfiles_flat(self, bucket: str, name: str) -> AsyncIterator[FileListEntry]:
+        assert name.endswith('/')
+        params = {
+            'prefix': name,
+            'delimiter': '/',
+            'includeTrailingDelimiter': 'true'
+        }
+        async for page in await self._storage_client.list_objects(bucket, params=params):
+            prefixes = page.get('prefixes')
+            if prefixes:
+                for prefix in prefixes:
+                    url = f'gs://{bucket}/{prefix}'
+                    if url.endswith('/'):
+                        url = url[:-1]
```

Comment:
So is the contract here that every GoogleStorageFileListEntry must not end in a '/'